File: dawn_ocaml/lib/link_cut_tree.ml
open Base

module LinkCutTree = struct
  type node = {
    value: int;
    parent: node option ref;
    left: node option ref;
    right: node option ref;
    mutable rev: bool;
    mutable sum: int;
  }

  let make_node value =
    {
      value;
      parent = ref None;
      left = ref None;
      right = ref None;
      rev = false;
      sum = value;
    }

  let get_sum node =
    match !node with
    | None -> 0
    | Some n -> n.sum

  let update_sum node =
    match !node with
    | None -> ()
    | Some n ->
        n.sum <-
          n.value + get_sum n.left + get_sum n.right

  let push node =
    match !node with
    | None -> ()
    | Some n ->
        if n.rev then (
          n.rev <- false;
          let temp = n.left in
          n.left <- n.right;
          n.right <- temp;
          Option.iter
            (fun left ->
              left.rev <- not left.rev)
            !n.left;
          Option.iter
            (fun right ->
              right.rev <- not right.rev)
            !n.right)

  let make_root node =
    let rec loop x =
      push x;
      match !x.parent with
      | None -> ()
      | Some y ->
          y.left := None;
          y.right := None;
          x.parent := None;
          loop (Some y)
    in
    loop node

  let splay node =
    let rec zig y =
      match !y.parent with
      | None -> make_root y
      | Some x ->
          if !x.left == Some y then (
            push x;
            push y;
            let p = !x.parent in
            x.parent := Some (!y.parent);
            y.parent := p;
            if p == None then make_root y
            else if Some (!p).left == Some x then
              p.left := Some y
            else p.right := Some y;
            x.left := !y.right;
            y.right := Some x;
            update_sum x;
            update_sum y)
          else zig x
    and zig_zig y =
      match !y.parent with
      | None -> make_root y
      | Some x ->
          zig_zig x;
          zig y
    and zig_zag y =
      let x = Option.get !y.parent in
      push x;
      push y;
      let p = !x.parent in
      let q = !y.right in
      x.parent := Some (!q.parent);
      y.parent := Some (!q.parent);
      q.parent := Some x;
      x.left := Some q;
      y.right := !q.left;
      q.left := Some y;
      if p == None then make_root q
      else if Some (!p).left == Some x then
        p.left := Some q
      else p.right := Some q;
      update_sum x;
      update_sum y;
      update_sum q
    in
    make_root node;
    zig_zag node;
    update_sum node

  let access node =
    splay node;
    let rec loop x rev =
      if !x == node then rev
      else
        (push x;
         let rev' =
           (match !x.left with
            | None -> not rev
            | Some y -> loop y rev)
         in
         x.rev <- rev';
         update_sum x;
         rev')
    in
    ignore (loop node false)

  let link (x, y) =
    access x;
    access y;
    x.parent := Some y;
    update_sum y

  let cut x =
    access x;
    match !x.left with
    | None -> ()
    | Some y ->
        y.parent := None;
        x.left := None;
        update_sum x

  let root x = access x; !x.parent = None

  let lca (x, y) =
    access x;
    let x_sum = get_sum x in
    access y;
    let rec loop z =
      let z_sum = get_sum z in
      if z_sum >= x_sum && z_sum >= get_sum y then z
      else
        (access (Option.get !z.parent);
         loop (Option.get !z.parent))
    in
    loop y

  let path_sum (x, y) =
    let z = lca (x, y) in
    access x;
    access y;
    get_sum x + get_sum y - 2. *. get_sum z
end

File: dawn_rust/src/link_cut_tree.rs
pub struct Node {
    value: i32,
    parent: Option<*mut Node>,
    left: Option<*mut Node>,
    right: Option<*mut Node>,
    rev: bool,
    sum: i32,
}

impl Node {
    pub fn new(value: i32) -> Self {
        Node {
            value,
            parent: None,
            left: None,
            right: None,
            rev: false,
            sum: value,
        }
    }

    pub fn get_sum(&self) -> i32 {
        self.sum
    }

    pub fn update_sum(&mut self) {
        self.sum = self.value
            + self.left.as_ref().map_or(0, |n| unsafe { &**n }.get_sum())
            + self.right.as_ref().map_or(0, |n| unsafe { &**n }.get_sum());
    }

    pub fn push(&mut self) {
        if self.rev {
            self.rev = false;
            std::mem::swap(&mut self.left, &mut self.right);
            if let Some(l) = self.left.as_mut() {
                unsafe { &mut **l }.rev = !unsafe { &mut **l }.rev;
                unsafe { &mut **l }.push();
            }
            if let Some(r) = self.right.as_mut() {
                unsafe { &mut **r }.rev = !unsafe { &mut **r }.rev;
                unsafe { &mut **r }.push();
            }
        }
    }

    pub fn make_root(&mut self) {
        self.push();
        if let Some(p) = self.parent.as_mut() {
            unsafe { &mut **p }.left = None;
            unsafe { &mut **p }.right = None;
            unsafe { &mut **p }.make_root();
            self.parent = None;
        }
    }

    pub fn splay(&mut self) {
        self.make_root();
        let mut p = self.parent;
        while let Some(pp) = p {
            let mut pp_ptr = pp as *mut Node;
            let p_ptr = p as *mut Node;
            unsafe { &mut *p_ptr }.make_root();
            if let Some(ppp) = unsafe { &mut *p_ptr }.parent {
                let ppp_ptr = ppp as *mut Node;
                unsafe { &mut *ppp_ptr }.make_root();
                if (unsafe { &*ppp_ptr }.left.as_ref() == Some(&p_ptr))
                    == (unsafe { &*p_ptr }.left.as_ref() == Some(self))
                {
                    unsafe { &mut *p_ptr }.rotate();
                } else {
                    self.rotate();
                }
            }
            self.rotate();
            p = self.parent;
        }
        self.update_sum();
    }

    fn rotate(&mut self) {
        let p = unsafe { &mut *self.parent.unwrap() };
        let pp = p.parent;
        if let Some(pp_ptr) = pp {
            let pp_mut = unsafe { &mut *pp_ptr };
            if pp_mut.left.as_ref() == Some(p) {
                pp_mut.left = Some(self);
            } else {
                pp_mut.right = Some(self);
            }
        }
        self.parent = pp;
        if p.left.as_ref() == Some(self) {
            p.left = self.right;
            self.right = Some(p);
        } else {
            p.right = self.left;
            self.left = Some(p);
        }
        p.parent = Some(self);
        p.update_sum();
        self.update_sum();
    }

    pub fn access(&mut self) {
        self.splay();
        self.rev = false;
        let mut c = self.left;
        while let Some(c_ptr) = c {
            let mut c_mut = unsafe { &mut *c_ptr };
            c_mut.splay();
            c_mut.rev = !c_mut.rev;
            c_mut.update_sum();
            c_mut.right = Some(self);
            self.parent = Some(c_mut);
            c = self.left;
        }
    }

    pub fn link(&mut self, other: &mut Node) {
        self.access();
        other.access();
        other.parent = Some(self);
        other.update_sum();
    }

    pub fn cut(&mut self) {
        self.access();
        if let Some(l) = self.left.take() {
            unsafe { &mut *l }.parent = None;
            self.update_sum();
        }
    }

    pub fn root(&self) -> bool {
        self.parent.is_none()
    }

    pub fn lca(&mut self, other: &mut Node) -> *mut Node {
        other.access();
        let other_sum = other.get_sum();
        self.access();
        let mut z = other;
        while unsafe { &*z }.get_sum() < other_sum || unsafe { &*z }.get_sum() < self.get_sum() {
            z.access();
            z = unsafe { &mut *z }.parent.unwrap();
        }
        z
    }

    pub fn path_sum(&mut self, other: &mut Node) -> i32 {
        let z = self.lca(other);
        unsafe { &mut *self }.access();
        unsafe { &mut *other }.access();
        let x_sum = self.get_sum();
        let y_sum = other.get_sum();
        let z_sum = unsafe { &*z }.get_sum();
        x_sum + y_sum - 2 * z_sum
    }
}